### ISSUE TITLE
suggest replacing if comparison chain with `cmp` match

### ISSUE THREAD

**Issue by oli-cosmian** [rust-lang/rust-clippy#4531]
If one writes

```rust
if x < y {
     ...
} else if x > y {
     ...
} else {
     ...
}
```

one should instead be writing

```rust
match x.cmp(&y) {
     Ordering::Less => ...,
     Ordering::Greater => ...,
     Ordering::Equal => ...,
}
```

**Comment by james9909** [comment]
I can take a stab at this issue

**Comment by flip1995** [comment]
Thanks! Let us know if you have any questions.

**Comment by untitaker** [comment]
This requires an additional import and is harder to read. Why is this a default lint?

**Comment by oli-cosmian** [comment]
I don't think it's harder to read. imo this is much clearer what's going on, because you have the three cases very explicitly.

Also I don't consider additional or fewer imports to be relevant for whether a lint is useful or not

**Comment by untitaker** [comment]
In my book `a < b` will always be easier to read than `a.cmp(b) == Ordering::Less`. Irrespective of whether your proposed coding style provides more compile-time safety than a regular comparison, because that has nothing to do with readability.

Why would I import more traits and enums from std if I can achieve the same thing without? e.g. I don't import `FromStr` in favor of using `.parse()`

**Comment by mitsuhiko** [comment]
In the case of symbolic in particular it’s especially weird because the type doesn’t even implement `Ord` but only `PartialOrd` which requires one to use `partial_cmp` which is extra ugly.

**Comment by oli-cosmian** [comment]
> FWIW this lint also triggers for code that does not intend to handle all three cases: 

That looks also fine to me. I think it should be warn by default.

Since some situations require `partial_cmp`, we should fix the lint of course.

To me this lint is a clear improvement (especially since it does help catch bugs).

**Comment by raphlinus** [comment]
FWIW, I dislike this lint, and find the direct comparisons easier to read. As a result, https://github.com/xi-editor/druid/pull/406 is adding `#[allow(clippy:comparison_chain)]`.

**Comment by llogiq** [comment]
@raphlinus that's OK. Readability is not something that can be decided in the general case, but must be approached on a case-by-case basis.

**Comment by untitaker** [comment]
Cool, so you agree such a general lint is wrong?

**Comment by flip1995** [comment]
A check for `Ord` was implemented in #4827. Maybe it didn't hit stable/beta/nightly (Not sure where we're in the release cycle, since I was on vacation the last week) yet.

> Cool, so you agree such a general lint in defaults is wrong?

This is a `style` lint, which is the "weakest" default group and is expected to has the most `allows`. I also agree, that an exhaustive match is more rusty, than an if with comparisons. 

> this lint also triggers for code that does not intend to handle all three cases

See #4725
